[WASI] preview2 plugins, resource type instantiation, and related types export from plugin

dannypsnl opened the pull request:
### preview2 interfaces

- provide structure 2024/09
- provide content 2024/10-
    - refers to #2939 to get more details

- wasi:io/error
- wasi:io/streams
- wasi:cli/environment
- wasi:cli/exit
- wasi:cli/stdin
- wasi:cli/stdout
- wasi:cli/stderr
- wasi:clocks/wall-clock
- wasi:filesystem/types
- wasi:filesystem/preopens
- wasi:http/types

### What else provided in this PR

1. fix resource type internal fields 2024/10
2. fix component exported core module behaviour, it should not always require to expose host functions 2024/10
3. host function now works with `u8` 2024/10
4. `resource.drop` 2024/10-11
5. internal representation of types 2024/11

### TODO here
- [x] 2024/10/15 core module import section instantiation will crash (reason: when `FunctionInstance` is not a host function, I insert a deftype for it but no owner for the deftype)
- [x] 2024/10/17 provide resource implementation, one can code as below
    ```cxx
    class IoError : ResourceType {
    public:
      IoError() : ResourceType(new ComponentInstance("error")) {
        Impl->addHostFunc("...", ...);
        // ...
      }
    };
    ```
- [x] 2024/10-11 complete all resources content that standard required (at least to make example run)
- [x] 2024/11 fix record has no owner problem, currently, below code will not insert `Datetime` properly
    ```cxx
    class Datetime : public Record {
    public:
      Datetime()
          : Record({LabelValType("seconds", PrimValType::U64),
                    LabelValType("nanoseconds", PrimValType::U32)}) {}
    };

    WallClockModule::WallClockModule()
        : ComponentInstance("wasi:clocks/wall-clock@0.2.0") {
      addExport("datetime", Datetime());
    }
    ```

    This is a weird, since `addExport` will use right value `DefType&&` and `move` the type, then insert into the internal map of component instance. In `Datetime` case, which is a `Record`, the `variant` hierarchy is `Record <: DefValType <: DefType`, so I guess there might some problems in `variant`, if I use it then has to use the wrapping stack explicitly.

- [x] 2025/1-2025/2 refactoring `InterfaceType`
- [x] 2025/3 implements new resource machinery
- [x] 2025/3 instantiation: with function, with type
- [x] 2025/3 import type bound
- [ ] 2024/12-2025/ wasi:http

Commit:
[Preview2] plugins

+ wasi-io/error
+ wasi:io/streams

+ wasi-cli/environment
+ wasi:cli/exit
+ wasi:cli/stdin
+ wasi:cli/stdout
+ wasi:cli/stderr

+ wasi:filesystem/types
+ wasi:filesystem/preopens

+ wasi:clocks/wall-clock

[Component] named type recording

draft `SortCase::Type`
common `CInst` extraction
export types needed here
Now we can try to solve missing functions part

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[WASI] wasi:clock export datetime (a record type)

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] correct resource type concept

* add impl field in ResourceType
* add hint about default value of fields

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Fix] put needed constructor back

remove repeated
note undone part

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] module export core functions correctly

Component can instantiate a module and exports some core functions from
many places, previously, we only face the case those are host functions,
which leads to a wrong restriction been put.

Therefore, in this version I remove that, and by putting log pre/post, I
get

[2024-10-08 09:47:50.389] [info] export fd_write
[2024-10-08 09:47:50.389] [info] export fd_write complete!
[2024-10-08 09:47:50.389] [info] export environ_get
[2024-10-08 09:47:50.389] [info] export environ_get complete!
[2024-10-08 09:47:50.389] [info] export environ_sizes_get
[2024-10-08 09:47:50.389] [info] export environ_sizes_get complete!
[2024-10-08 09:47:50.389] [info] export proc_exit
[2024-10-08 09:47:50.389] [info] export proc_exit complete!

which is desired!

* rename internal function of component instance

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] fix export function in export module

The subtype here is created, so need someone owned it, here I pick the
export module itself. In this case, the module is created by component,
as the interface of the component.

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Misc] move comment

remove debug print

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] draft form of resource type in plugin

Also, print out resource type in instruction to see what we got here.
modify all resources, also tries to fix datetime missing error
update comment
note about todo in Record value
drafting resource.drop instruction

resource.drop is created for invoking drop function recorded in
ResourceType, it's lowering signature is
  [i32] -> []
when sync, and is
  [i32] -> [i32]
if async

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Internal] improve format output of component types

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] exploring how to provide dtor of resource from plugin

From binary we can see code like

  import [func 2] Import { module: "wasi:io/streams@0.2.0", name:
"[resource-drop]output-stream", ty: Func(0) }

so we can figure out that every resource related function is export as
an usual host-function

rename addExport(DefType) to addHostType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] complete alias type command

misc:
* remove a unused debug message
* comment out message for now
* get destructor of resource from component instance
* use visitor to break down complicated if-else for Canon commands
* let lowering and lifting of Executor been public, since visitor needs
  them
* use check style to ensure instruction is facing a ResourceType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] complete resource.drop instruction

* avoid shadowing
* drop functions for wasi-io
* add host functions for wasi:cli/environment
* implements wasi:cli/exit
  * add uint8_t (due to wasi:cli/exit)
  * hostfunc: exit
  * hostfunc: exit-with-code
* eof
* the todo seems not need now
* fix build
* refine export cases

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
wasi:io drop io error

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add methods of wasi:io/streams output-stream

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
place more methods

* [method]output-stream.blocking-write-and-flush
* [method]descriptor.write-via-stream

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:filesystem/preopens@0.2.0] get-directories

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
complete needed wasi:filesystem/types host functions

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
improve resource-type pretty-print

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Tuple

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
record what todo

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[WASI] fix signatures and note some unknown what todo signature

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Option

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Enum

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Misc] correct include

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Result

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] resource.drop is a runtime decision

Thus, no need to implement anything about it from wasi side

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] implements resource.drop idea

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] provide type for generated resource drop function

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[WASI] fix signatures that will fail the instantiation

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] core table alias

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
ignore string encoding options

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] rephrase flatten types process

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Record

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] flatten type for record

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] ValComp, add Variant

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
extract recordty's formatter

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
extract formatter

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
use initializer list

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
correct the addHostType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
extract runtime instance component part

This extraction improves the build time, since component instance is
often get modified at this moment

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
continue extraction

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
draft of wasi:http/types

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
complete `variant method`

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add host type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
scheme

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
error-code

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
eof

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add method

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
remove comment that mentions a fixed problem

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
correct wasi-io/stream stream-error

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
start draft of stream-error

An important lesson here is: separate Variant and VariantTy as much as
possible, to give a proper constructor for both is impossible, and also
not quite make sense except that it's convenience.

This implementation raises an issue: the construction functions of
StreamError lead symbol duplicates, needs to clean this plugin, to get a
better organization

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
solve symbol duplicates

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
use namespace method

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
note about TODO

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
field-key and header-error

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add type field-value and fixing cmake setup

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
fix several misc error in wasi-http

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
fix Option<T>

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
to correctly store higher type like option<T>, need to convert them into index

The problem is, I cannot sure I should reuse index for the same type or
not; and if I use these type, e.g. ErrorCode, how can I ensure use-site
will lookup at the right place (wasi-http plugin component), instead of
explain these type indicies at their place (use-site component)?

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
more resource

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
status-code

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
complete wasi:http/types surface

NOTE that still need to fill resource

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
flatten target should be an argument

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
resource `fields` constructor

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add todolist

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
convert `getType` of component instance to `Expect<T>`

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
getCoreType and getCoreGlobalInstance now forward array outbound

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
protect core memory instance access

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
improve core-table and core-function access

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
protect module instance access

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
protect value access

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
protect component instance access

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
complete more ValType to ValueType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Component] support all integer component model have

* [wasi-http] incoming-response [method] status
  * provide a proper default value

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi-http] incoming-response.headers

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] remove the wrong s8-s64

1. use existed I8, my fault is forgot the I32 and I64 is handled, only
I8 and I16 needs some conversions
2. fix constructors and methods type accrodings the type mismatching
report
3. wasi-http [constructor]outgoing-response

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi-http] more placeholders

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi-http] add [static]outgoing-body.finish

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[canonical] implements more flatten type

1. implements discriminant_type
2. implements flatten_result (modify flatten_variant)
   the next of this should be invoke flatten_variant with proper
arguments to avoid repeated copy the algorithm
3. correct flattenFunctionType in lower mode, the next should be correct
lift part so we always use flatten algorithm

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[Internal] canonical options can be printout

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
InterfaceType now works in a parallel structure

parallel from ValType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
remove unused message

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[test] fix build

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
fix rebased build failed

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[linter] fix typo in include/ast/component/canonical.h

unkown -> unknown

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[fix] use more standard type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
explicit use ValType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
static cast type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[msvc] avoid shadowing

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
note about the special case

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
despecialize & improve InterfaceType

InterfaceType now supports
1. case name to represent variant case
2. `bool isNone()` for flatten algorithm

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
flatten variant

and join helper

NOTE: the output type still wrong, I shall check another version of tool
Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[canon] handle numberic types lowering

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[canon] use direct style to port algorithm

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:io] add missing self

TODO: In fact, shall use `Borrow<T>`, but for now this is enough to pass
type checking

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
add context

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:http] set `response-outparam.set` type correct

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] resource table management

1. update type of `Resources` in `ComponentInstance`
2. add `FreeResources` to compute next index when new a thing
3. update `removeResource` machinery
4. remember `this` as the component instance that `Env` should know, so
later host functions in plugin can operate resources

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
fix build and hint a problem

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:http] `outgoing-response.body` type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:http] `outgoing-body.write` type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[wasi:http] `outgoing-body.finish` type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

juntao commented:
Hello, I am a [code review agent](https://github.com/flows-network/github-pr-review/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

dannypsnl commented:
NOTE: Rename branch, replace #3753.

Commit:
[component] lift more types

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] lift primitive value type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] lift own type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] remove unused log

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] improve comment about InterfaceType

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] instantiation arguments

* with function
* with type

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] import type bound

The algorithm is basically
1. lookup a component by IMPORT NAME
2. check the component has expected type, type bound has two cases
  - eq i: Lookup type `i` and check type are the same
  - sub resource: Ensure the imported component is a resource is enough
(though our internal representation of component instance cannot do this
yet)

For now, I leave type check behind. The bigger problem is import cannot
immediately tries to `findComponent` by name, because this binding
wasn't inserted into `StoreMgr` yet, we need a way to delay lookup

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] adjust instantiation code

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

dannypsnl commented on plugins/wasi_io/env.h:
```diff
@@ -0,0 +1,21 @@
+
+class WasiIOEnvironment {
+public:
+  std::vector<OutputStream> OutputStreamList;
```
We don't have to manage resource in plugin.

Commit:
[misc] add modifier

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] use EXPECTED_TRY to reduce code size

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] add const modifier

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[misc] update error message

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] postpone import statement execution

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] protect function instance access

* provide better error message
* update related invoke

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

Commit:
[component] correct the instantiate component instance logic

The correct concept is
1. finding a component (class)
2. instantiate an instance via the component just got
3. put arguments to proper place
4. `executeImports` for the instance to setup the instance
5. insert the instance back the current (component) instance

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

dannypsnl commented:
update, current error message is

```
[2025-03-25 17:03:05.129] [error] component: ``, access function instance: 0/0
[2025-03-25 17:03:05.129] [error] while exporting component function `handle`.
```

This is because an export section of a component also happened before instantiation, and it is very funny that hello-wasi-http example has an export that uses an imported function in the final component.

Commit:
[component] defer export & fix instance Type case

When instantiate argument has kind Type, it seems still a component
instance, therefore, we should put it into StoreMgr

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@secondstate.io>

dannypsnl commented:
Sort of,
1. need to fix build on different platforms
2. there are some concepts need to be carefully re-understood and rewrote, but they can be another PR

q82419 commented on plugins/wasi_cli/base.h:
```diff
@@ -0,0 +1,25 @@
+namespace Host {
+
+template <typename T>
+class WasiCli : public Runtime::Component::HostFunction<T> {
```
Comment:
Maybe add a namespace `Preview2`.
Since there will be preview 3 in the future.

dannypsnl replied:
Agree

q82419 commented on plugins/wasi_cli/env.h:
```diff
@@ -0,0 +1,16 @@
+namespace WasmEdge {
+namespace Host {
+
+class WasiCliEnvironment {};
```
If there's no necessary shared data, the environment class is not needed.

q82419 commented on lib/runtime/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: 2019-2024 Second State INC
+
+wasmedge_add_library(wasmedgeRuntime
```
We supposed that `wasmedgeRuntime` to be a header-only library.
Would you consider this?

q82419 reviewed (commented):
I've viewed the plugins and runtime.

1. Seems like the preview 2 functions are empty.
2. Seems like the `wasi-http` only added the http-types. The host functions are not connected.
3. Please consider the `wasmedgeRuntime` to be a header-only library.

dannypsnl commented:
> 2. Seems like the `wasi-http` only added the http-types. The host functions are not connected.

Because wasi-http program should provide callback, which use types in wasi-http.

q82419 commented:
I'll squash and rebase the commits to the branch `proposal/component`.
Due to some unknown issues that causes CI fail, I'll close this PR after the codes ready on that branch.

q82419 commented:
Comment:

I've squashed this PR and rebased onto the branches `proposal/component` and `proposal/wasi_preview2`.
I'll keep extract the partitions of that branches when refactoring.
Keep this PR for code tracing.

q82419 commented:
@dannypsnl 

Can you help to explain why the [code here](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/component/component_type.cpp#L494-L500):

The `Loader::loadExportDecl` will load the `core:exportdecl`, which is:
```
core:exportdecl  ::= n:<core:name> d:<core:importdesc>                    => (export n d)
```
Where the `core:name` is a utf-8 bytes vector.

But the code above use the [`Loader::loadExportName`](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/component/component_type.cpp#L367-L380), seems like it has a `0x00` prefix byte.

q82419 commented:
@dannypsnl 

You implemented a new `AST::Component::FunctionType` for the component function and host function.
With lifting/lowering to `Component::FuncType`, as the [binary definition](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Binary.md#type-definitions), the functype is:

```
functype      ::= 0x40 ps:<paramlist> rs:<resultlist>     => (func ps rs)
paramlist     ::= lt*:vec(<labelvaltype>)                 => (param lt)*
resultlist    ::= 0x00 t:<valtype>                        => (result t)
                | 0x01 0x00
```

Seems like it only accepts at most one result value.
But in [loader](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/component/component_type.cpp#L264-L268), you loaded the value type vector in `0x01` case of result list.
Is it the wrong implementation?

q82419 commented:
Due to far away from current component model implementation, close this.
